File: solution/1110. Delete Nodes And Return Forest/1110.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        for (int d : to_delete)
            D.insert(d);

        dfs(root, true);
        return res;
    }
    
private:
    vector<TreeNode*> res;
    unordered_set<int> D;
    
    TreeNode* dfs(TreeNode* node, bool is_root) {
        if (!node) return nullptr;
        
        bool deleted = D.count(node->val);
        if (is_root && !deleted) res.push_back(node);
        
        node->left = dfs(node->left, deleted);
        node->right = dfs(node->right, deleted);
        
        return deleted ? nullptr : node;
    }
};

File: solution/1110. Delete Nodes And Return Forest/1110.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def delNodes(self, root: Optional[TreeNode], to_delete: List[int]) -> List[TreeNode]:
        D = set(to_delete)
        res = []
        
        def dfs(node, is_root):
            if not node: return None
            
            deleted = node.val in D
            
            if is_root and not deleted:
                res.append(node)
            
            node.left = dfs(node.left, deleted)
            node.right = dfs(node.right, deleted)
            
            return None if deleted else node
        
        dfs(root, True)
        return res
    
